src/engine/IdTable.h

```diff
@@ -0,0 +1,933 @@
+  }
+  const Id& operator()(size_t row, size_t col) const {
+    return IdTableImpl<COLS>::_data[row * IdTableImpl<COLS>::_cols + col];
+  }
+
+  // Row access
+  Row operator[](size_t row) {
+    return Row(IdTableImpl<COLS>::_data + row * IdTableImpl<COLS>::_cols,
+               IdTableImpl<COLS>::_cols);
+  }
+
+  ConstRow operator[](size_t row) const {
+    // Moving this method to impl allows for efficient ConstRow types when
+    // using non dynamic IdTables.
+    return IdTableImpl<COLS>::getConstRow(row);
+  }
+
+  // Begin iterator
+  iterator begin() {
+    return iterator(IdTableImpl<COLS>::_data, 0, IdTableImpl<COLS>::_cols);
+  }
+  const_iterator begin() const {
+    return iterator(IdTableImpl<COLS>::_data, 0, IdTableImpl<COLS>::_cols);
+  }
+  const_iterator cbegin() const {
+    return iterator(IdTableImpl<COLS>::_data, 0, IdTableImpl<COLS>::_cols);
+  }
+
+  // End iterator
+  iterator end() {
+    return iterator(IdTableImpl<COLS>::_data, IdTableImpl<COLS>::_size,
+                    IdTableImpl<COLS>::_cols);
+  }
+  const_iterator end() const {
+    return iterator(IdTableImpl<COLS>::_data, IdTableImpl<COLS>::_size,
+                    IdTableImpl<COLS>::_cols);
+  }
+  const_iterator cend() const {
+    return iterator(IdTableImpl<COLS>::_data, IdTableImpl<COLS>::_size,
+                    IdTableImpl<COLS>::_cols);
+  }
+
+  Row back() {
+    return Row(IdTableImpl<COLS>::_data +
+                   (IdTableImpl<COLS>::_size - 1) * IdTableImpl<COLS>::_cols,
+               IdTableImpl<COLS>::_cols);
+  }
+
+  ConstRow back() const {
+    return ConstRow(IdTableImpl<COLS>::_data + (IdTableImpl<COLS>::_size - 1) *
+                                                   IdTableImpl<COLS>::_cols,
+                    IdTableImpl<COLS>::_cols);
+  }
+
+  Id* data() { return IdTableImpl<COLS>::_data; }
+  const Id* data() const { return IdTableImpl<COLS>::_data; }
+
+  // Size access
+  size_t rows() const { return IdTableImpl<COLS>::_size; }
+
+  /**
+   * The template parameter here is used to allow for creating
+   * a version of the function that is constexpr for COLS != 0
+   * and one that is not constexpt for COLS == 0
+   */
+  using IdTableImpl<COLS>::cols;
+
+  size_t size() const { return IdTableImpl<COLS>::_size; }
+
+  /**
+   * @brief Sets the number of columns. Should only be called while data is
+   *        still nullptr.
+   **/
+  void setCols(size_t cols) {
+    assert(IdTableImpl<COLS>::_data == nullptr);
+    IdTableImpl<COLS>::setCols(cols);
+  }
+
+  void emplace_back() { push_back(); }
+
+  void push_back() {
+    if (IdTableImpl<COLS>::_size + 1 >= IdTableImpl<COLS>::_capacity) {
+      grow();
+    }
+    IdTableImpl<COLS>::_size++;
+  }
+
+  void push_back(const std::initializer_list<Id>& init) {
+    assert(init.size() == IdTableImpl<COLS>::_cols);
+    if (IdTableImpl<COLS>::_size + 1 >= IdTableImpl<COLS>::_capacity) {
+      grow();
+    }
+    std::memcpy(IdTableImpl<COLS>::_data +
```

Comment:
I think if we are using `realloc` we need to check if it did reuse the memory in which case we can skip this `memcpy`

Reply:
Never mind I was thinking of another `memcpy` didn't realize this was just for the `init`